models/events.go

```diff
@@ -1,114 +1,142 @@
+// collection. It also performs safety checks on the provided event data.
+func decodeCadenceEvents(events flow.BlockEvents) (*CadenceEvents, error) {
+	e := &CadenceEvents{events: events}
 
-	cumulativeGasUsed := uint64(0)
-	var lastReceipt *StorageReceipt
-	for _, e := range c.events.Events {
-		if isTransactionExecutedEvent(e.Value) {
-			tx, receipt, err := decodeTransactionEvent(e.Value)
-			if err != nil {
-				return nil, nil, err
+	// decode and cache block and transactions
+	for _, event := range events.Events {
+		val := event.Value
+
+		if isBlockExecutedEvent(val) {
+			if e.block != nil { // safety check, we can only have 1 or 0 evm blocks per one flow block
+				return nil, fmt.Errorf("EVM block was already set for this Flow block, invalid event data")
 			}
 
-			if lastReceipt != nil && lastReceipt.BlockNumber.Cmp(receipt.BlockNumber) == 0 {
-				cumulativeGasUsed += lastReceipt.GasUsed
-				receipt.CumulativeGasUsed = cumulativeGasUsed
-			} else {
-				cumulativeGasUsed = receipt.GasUsed
-				receipt.CumulativeGasUsed = cumulativeGasUsed
+			block, err := decodeBlockEvent(val)
+			if err != nil {
+				return nil, err
 			}
 
-			lastReceipt = receipt
+			e.block = block
+			continue
+		}
 
-			txs = append(txs, tx)
-			rcps = append(rcps, receipt)
+		if isTransactionExecutedEvent(val) {
+			tx, receipt, err := decodeTransactionEvent(val)
+			if err != nil {
+				return nil, err
+			}
+
+			e.transactions = append(e.transactions, tx)
+			e.receipts = append(e.receipts, receipt)
 		}
 	}
 
-	return txs, rcps, nil
+	// safety check, we can't have an empty block with transactions
+	if e.block == nil && len(e.transactions) > 0 {
+		return nil, fmt.Errorf("EVM block can not be nil if transactions are present, invalid event data")
+	}
+
+	return e, nil
 }
 
-// Empty checks if there are any evm block or transactions events.
-// If there are no evm block or transactions events this is a heartbeat
-// event that is broadcast in intervals.
-func (c *CadenceEvents) Empty() bool {
-	for _, e := range c.events.Events {
-		if isTransactionExecutedEvent(e.Value) || isBlockExecutedEvent(e.Value) {
-			return false
-		}
-	}
+// Block evm block. If event doesn't contain EVM block the return value is nil.
+func (c *CadenceEvents) Block() *Block {
+	return c.block
+}
+
+// Transactions included in the EVM block, if event doesn't
+// contain EVM transactions the return value is nil.
+func (c *CadenceEvents) Transactions() []Transaction {
+	return c.transactions
+}
 
-	return true
+// Receipts included in the EVM block, if event doesn't
+// contain EVM transactions the return value is nil.
+func (c *CadenceEvents) Receipts() []*StorageReceipt {
+	return c.receipts
+}
+
+// Empty checks if there is an EVM block included in the events.
+// If there are no evm block or transactions events this is a heartbeat event.
+func (c *CadenceEvents) Empty() bool {
+	return c.block == nil
```

Comment:
Theoretically, the system chunk transaction is run for each Flow block formation. Which means for each Flow block, there will be emitted an `EVM.BlockExecuted` event. I don't see how the EVM block will be `nil`, for responses from the Event Streaming API. Basically, we should get a response for every single Flow block. I believe `Empty` here means that the EVM block had no EVM transactions.